Geomap: Add dynamic initial view options

drew08t opened the pull request:
What this PR does / why we need it:
For dynamic data that changes, setting up map view to "fit data layers" does not update upon data change. As an extension to this support, minimum control options should also be included:

- [x] - Max zoom
- [x] - Padding (%)
- [x] - Last data point only (boolean)
- [x] - Handle single layer bug

After:
![Aug-31-2022 13-06-22](https://user-images.githubusercontent.com/60050885/187772219-e8eb6c89-2a94-4450-872c-b89c79e00415.gif)

Closes #40496

drew08t committed:
Geomap: Add dynamic initial view options

drew08t committed:
Add contol options for dynamic initial view

drew08t committed:
Add handling for last only scope

drew08t commented on public/app/plugins/panel/geomap/GeomapPanel.tsx:
```diff
@@ -592,19 +594,28 @@ export class GeomapPanel extends Component<Props, State> {
   }
 
   initViewExtent(view: View, config: MapViewConfig, layers: Collection<BaseLayer>) {
+    //TODO add option to set extent based on last data point only
```
remove this.

drew08t commented on public/app/plugins/panel/geomap/GeomapPanel.tsx:
```diff
@@ -592,19 +594,28 @@ export class GeomapPanel extends Component<Props, State> {
-          const extent = getLayersExtent(layers);
+          const extent = getLayersExtent(layers, config.lastOnly ?? false);
           if (!isEmpty(extent)) {
+            //TODO make padding an option
```
remove this.

drew08t commented on public/app/plugins/panel/geomap/GeomapPanel.tsx:
```diff
@@ -267,6 +267,8 @@ export class GeomapPanel extends Component<Props, State> {
       }
     }
+    // Because data changed, check map view and change if needed (data fit)
+    this.map!.setView(this.initMapView(this.props.options.view, this.map!.getLayers()));
```
Should add conditional for data fit, so this view init only gets called for that view setting.

drew08t committed:
Remove stale todos

drew08t committed:
Only reinitialize map view during data if fit

nmarrs commented on public/app/plugins/panel/geomap/editor/MapViewEditor.tsx:
```diff
@@ -112,6 +112,37 @@ export const MapViewEditor: FC<StandardEditorProps<MapViewConfig, any, GeomapPan
       </InlineFieldRow>
 
+      {value?.id === MapCenterID.Fit && (
+        <>
```
This would probably make sense to split out into an individual component for improved maintainability / testing

drew08t replied:
Will look into that after resolving conflicts on GeomapPanel.

nmarrs commented on public/app/plugins/panel/geomap/editor/MapViewEditor.tsx:
```diff
@@ -112,6 +112,37 @@ export const MapViewEditor: FC<StandardEditorProps<MapViewConfig, any, GeomapPan
+                value={value?.lastOnly ? 'last' : 'all'}
+                options={[
+                  { label: 'All', value: 'all' },
+                  { label: 'Last Only', value: 'last' },
```
For values it might make sense to create an enum type for better maintainability

drew08t replied:
Agreed.

nmarrs commented on public/app/plugins/panel/geomap/GeomapPanel.tsx:
```diff
@@ -599,12 +604,19 @@ export class GeomapPanel extends Component<Props, State> {
-              maxZoom: config.zoom ?? config.maxZoom,
+              maxZoom: maxZoom,
             });
+            view.setResolution(res * (padding / 100 + 1));
```
any insight on this calculation? (i.e. why padding / 100 + 1)

drew08t replied:
Padding is a percentage > 0, so this is a multiplier to adjust the resolution accordingly. 50% for example would multiply resolution by 1.5 = 50 / 100 + 1.

nmarrs commented on public/app/plugins/panel/geomap/GeomapPanel.tsx:
```diff
@@ -599,12 +604,19 @@ export class GeomapPanel extends Component<Props, State> {
           coord = [config.lon ?? 0, config.lat ?? 0];
         } else if (v.id === MapCenterID.Fit) {
-          const extent = getLayersExtent(layers);
+          const extent = getLayersExtent(layers, config.lastOnly ?? false);
```
We could also just pass `config.lastOnly` here and set the default param to `false` in `getLayerExtent` definition
```
export function getLayersExtent(layers: Collection<BaseLayer>, lastOnly: boolean = false): Extent {
```

drew08t committed:
Add options for data fit

In map init, remove dependency on layers input.

Add a boolean to map view config to handle all layers, with default
set to true. Also, add layer string to handle currently selected layer.

Change some verbage.

In map view editor, add controls for data extent options. Only display
layer selection when all layers is not selected.

Update layer extent function to handle all options permutations. When
all layers is selected, return all data for extent calculation. When
last only is selected, only return last data point in matching layer for
extent calculation. Otherwise, return all data points in matching layer
for extent calculation.

drew08t commented:
New options controls for data fit:
![Sep-14-2022 09-46-28](https://user-images.githubusercontent.com/60050885/190214331-1a2db765-7461-49f2-99bc-f24a0e29fdab.gif)

drew08t committed:
Change padding tooltip and hide for last only

drew08t committed:
Simplify getLayersExtent call

drew08t committed:
Add enums for data scope options

drew08t committed:
Update GeomapPanel for refactor merge

drew08t committed:
Merge branch 'main' into drew08t/geomap-initial-view-options

drew08t committed:
Move view init functions back into geomap class

Re-apply data change handling and extent calculation handling.

drew08t committed:
Update padding tooltip verbage

drew08t committed:
Ensure geomap panel options layers are initialized

nmarrs committed:
Merge remote-tracking branch 'origin' into drew08t/geomap-initial-view-options

nmarrs committed:
Clean up GeomapPanel file (betterer)

nmarrs commented on public/app/plugins/panel/geomap/editor/MapViewEditor.tsx:
```diff
@@ -94,9 +130,69 @@ export const MapViewEditor: FC<
           </InlineFieldRow>
         </>
       )}
+      {value?.id === MapCenterID.Fit && (
```
It might be nice to break up this component up into separately defined fragments for better readability / maintainability (not just additions here but the entire component)

nmarrs reviewed (commented):
Overall this is looking great! I think we are getting close

A few things I noticed:
- We do not update the view when when switching layer types - maybe this isn't necessary though given we only really care about the underlying data updating -> though given current behavior with geojson / heatmap layers this may be a nice addition depending on where we land there
- Streaming data doesn't seem to work -> though that may be due to other issues (table view doesn't seem to work when streaming so there is no data to fit to?)
- This feature doesn't seem to work as expected for some layer types (heatmap, geojson) (see screencast below)


https://user-images.githubusercontent.com/22381771/191168102-6ce55f8c-0f50-486d-a145-b795604a43e6.mov

nmarrs committed:
Merge remote-tracking branch 'origin' into drew08t/geomap-initial-view-options

nmarrs commented:
Looked this over with Ryan and came to following conclusions:
For now heat map is a little hard to address given the nature of how heat map works, for now we can push this off

As for re-initializing view on layer type change, we can also hold off on that as well. Also, we probably will want to do some exploration for how initial view actually works (whether or not local storage is utilized) etc.

Streaming was a local issue -> It appears the `live-service-web-worker` feature flag seems to break streaming right now (current solution is to just disable / remove this feature flag and streaming works as expected)

nmarrs committed:
Merge remote-tracking branch 'origin' into drew08t/geomap-initial-view-options

nmarrs committed:
refactors / clean up

nmarrs committed:
more cleanup

nmarrs reviewed (approved):
This is looking good to me as a first implementation of this feature - I did make a number of refactors so it might be good if @drew08t / @ryantxu look them over to ensure I didn't miss something / test things locally to verify before merging in 👍 

Streaming even works :)

https://user-images.githubusercontent.com/22381771/191390973-4a8aa334-bfb9-4d13-8431-90d167303bd4.mov

drew08t commented:
> This is looking good to me as a first implementation of this feature - I did make a number of refactors so it might be good if @drew08t / @ryantxu look them over to ensure I didn't miss something / test things locally to verify before merging in 👍
> 
> Streaming even works :)
> 
>  Fit.data.streaming.demo.mov

That streaming video is awesome!

drew08t commented:
Refactoring and clean-up LGTM! Tested locally and works as expected!
